### PR TITLE
Dark core rgb pro

### PR DESCRIPTION
**Pull request by Clofriwen**
add support for Mice "Dark Core RGB Pro" (as experimental)

that should support Mice "Dark Core RGB Pro SE" (as experimental)

i did tests on mine (i have an SE version)

Place talk about it :
Issues : https://github.com/ckb-next/ckb-next/issues/716
Previous Pull asked : https://github.com/ckb-next/ckb-next/pull/840

Thanks to Lofty, KitsuWhooa and Gil @gilvbp for help.

### CODE REVIEW

**Comment by Clofriwen** [comment]
This mice is in experimental support.

To be able use it, you need to start daemon in experimental mode (as root) :
`ckb-next-daemon --enable-experimental` or `sudo ckb-next-daemon --enable-experimental`

you can have an autostart in experimental mode with SystemD by editing this file : 

> /etc/systemd/system/multi-user.target.wants/ckb-next-daemon.service

And change this
```
[Unit]
Description=Corsair Keyboards and Mice Daemon

[Service]
ExecStart=/usr/local/bin/ckb-next-daemon
Restart=on-failure

[Install]
WantedBy=multi-user.target
```
to it
```
[Unit]
Description=Corsair Keyboards and Mice Daemon

[Service]
Environment="CKB_ARG=--enable-experimental"
ExecStart=/usr/local/bin/ckb-next-daemon $CKB_ARG
Restart=on-failure

[Install]
WantedBy=multi-user.target
```
**### Warning :** don't do it if you don't know how to go back (like, from a Live CD)
Keep warm to path, that `/usr/bin/ckb-next-daemon` or `/usr/local/bin/ckb-next-daemon`

**Review by tatokis — CHANGES_REQUESTED** [review]
Other than the commented sections, it looks good to me.

Can you also build with -DCMAKE_BUILD_TYPE=Debug and make sure that the daemon does not throw warnings about duplicate lights?

Thanks!

**Comment by daugustin** [comment]
:partying_face: 

This now compiles.
I have not used CKB before so I'm not sure I'm using it right, but the DPI indicator lights always stay on the same colors. I would assume they would cycle with the assigned stages, but that does not seem to work. The different DPI stages work.

All other stuff seems to work.

Here is a startup with `-DCMAKE_BUILD_TYPE=Debug`

```
ckb-next: Corsair RGB driver 0.5.0-21-g7e60493
[I] Root controller ready at /dev/input/ckb0
[I] Adding device with vendor = 0x1b1c, product = 0x1b7f, path = /dev/bus/usb/001/003, syspath = /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7
[E] patchkeys (keymap_patch.c:151): Could not find profswitch in base keymap. Not patching profswitch for mouse.
[I] Connecting CORSAIR DARK CORE RGB PRO SE Gaming Dongle at /dev/input/ckb1
[I] Claiming interfaces. name=CORSAIR DARK CORE RGB PRO SE Gaming Dongle, firmware=0504, ep_str= 6
[I] ckb1: Claiming 6 interfaces
[I] Starting input thread for /dev/input/ckb1
[I] Found EP 0x82 at /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.0/ep_82
[I] Endpoint path /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.0/ep_82 has wMaxPacketSize 64
[I] Found EP 0x84 at /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.1/ep_84
[I] Endpoint path /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.1/ep_84 has wMaxPacketSize 64
[I] Found EP 0x83 at /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.2/ep_83
[I] Endpoint path /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.2/ep_83 has wMaxPacketSize 64
[I] Found EP 0x81 at /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.3/ep_81
[I] Endpoint path /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.3/ep_81 has wMaxPacketSize 64
[I] Found EP 0x85 at /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.4/ep_85
[I] Endpoint path /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.4/ep_85 has wMaxPacketSize 64
[I] Found EP 0x86 at /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.5/ep_86
[I] Endpoint path /sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7:1.5/ep_86 has wMaxPacketSize 64
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x15. Device said it's not supported.
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x16. Device said it's not supported.
[I] ckb1: Pairing id: b8 5e 19 13 04 2f 72 f6 
[I] Found new bragi subdevice 1
[I] ckb2: Subkb vendor: 0x1b1c, product: 0x1b7e
[E] patchkeys (keymap_patch.c:151): Could not find profswitch in base keymap. Not patching profswitch for mouse.
[I] Setup finished for /dev/input/ckb1
[W] bragi_process_notification (bragi_notification.c:36): ckb1: Unknown bragi notification for dev 2
[W] bragi_process_notification (bragi_notification.c:36): ckb1: Unknown bragi notification for dev 2
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x15. Device said it's not supported.
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x16. Device said it's not supported.
[I] ckb2: Pairing id: b8 5e 19 13 04 2f 72 f6 
[I] Setup finished for /dev/input/ckb2
[W] bragi_process_notification (bragi_notification.c:36): ckb1: Unknown bragi notification for dev 2
```

File: src/gui/keymap.h
Comment (Clofriwen) [reply src/gui/keymap.h:79]: Moved on 
```
        K57_WL,
        K63_WL,
        DARKCORERGBPRO,
        _MODEL_MAX
    };
```
in next commit

File: src/daemon/keymap_patch.c
Comment (Clofriwen) [reply src/daemon/keymap_patch.c:117]: Cleaned in next commit

File: src/daemon/led_bragi.c
Comment (Clofriwen) [reply src/daemon/led_bragi.c:39]: Cleaned in next commit

File: src/daemon/usb.c
Comment (Clofriwen) [reply src/daemon/usb.c:40]: Cleaned in next commit

File: src/daemon/usb.h
Comment (Clofriwen) [reply src/daemon/usb.h:154]: Cleaned in next commit

**Comment by Clofriwen** [comment]
Hello,
I have added some little thing :
- DPI 3 level indicator & Sniper mode (working like in official)
- Battery indicator (under wheel) (green = charging, orange = low, red = critical) (like in official... but not sure for orange)
- Profile switch now working
- Some little cleaning

The DPI level working better is that in black color and not in animation
DarkCoreRGBPro_DPI

Here, the newest build with -DCMAKE_BUILD_TYPE=Debug :

ckb-next-daemon :
```
ckb-next: Corsair RGB driver 0.5.0-unknown
[I] Root controller ready at /dev/input/ckb0
[I] Adding device with vendor = 0x1b1c, product = 0x1b7f, path = /dev/bus/usb/001/016, syspath = /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3
[E] patchkeys (keymap_patch.c:151): Could not find profswitch in base keymap. Not patching profswitch for mouse.
[I] Adding device with vendor = 0x1b1c, product = 0x1b20, path = /dev/bus/usb/001/003, syspath = /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-2
[I] Connecting CORSAIR DARK CORE RGB PRO SE Gaming Dongle at /dev/input/ckb1
[I] Claiming interfaces. name=CORSAIR DARK CORE RGB PRO SE Gaming Dongle, firmware=0204, ep_str= 5
[I] ckb1: Claiming 5 interfaces
[I] Connecting Corsair STRAFE RGB Gaming Keyboard at /dev/input/ckb2
[I] Claiming interfaces. name=Corsair STRAFE RGB Gaming Keyboard, firmware=0308, ep_str= 2
[I] ckb2: Claiming 2 interfaces
[I] Starting input thread for /dev/input/ckb1
[I] Starting input thread for /dev/input/ckb2
[I] Found EP 0x82 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.0/ep_82
[I] Found EP 0x81 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-2/1-2:1.0/ep_81
[I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-2/1-2:1.0/ep_81 has wMaxPacketSize 64
[I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.0/ep_82 has wMaxPacketSize 64
[I] Found EP 0x84 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.1/ep_84
[I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.1/ep_84 has wMaxPacketSize 64
[I] Found EP 0x83 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.2/ep_83
[I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.2/ep_83 has wMaxPacketSize 64
[I] Found EP 0x81 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.3/ep_81
[I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.3/ep_81 has wMaxPacketSize 64
[I] Found EP 0x85 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.4/ep_85
[I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.4/ep_85 has wMaxPacketSize 64
[I] Received identification packet: 0e 01 00 00 01 01 00 01 08 03 14 00 1c 1b 20 1b 01 01 01 08 c0 ff 40 01 03 00 00 00 02 44 01 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x15. Device said it's not supported.
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x16. Device said it's not supported.
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x96. Device said it's not supported.
[I] ckb1: Pairing id: ff ad 47 d6 bc 2b 6f 44 
[I] Found new bragi subdevice 1
[I] ckb3: Subkb vendor: 0x1b1c, product: 0x1b7e
[E] patchkeys (keymap_patch.c:151): Could not find profswitch in base keymap. Not patching profswitch for mouse.
[I] Setup finished for /dev/input/ckb1
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x15. Device said it's not supported.
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x16. Device said it's not supported.
[W] bragi_get_property (bragi_common.c:13): Failed to get property 0x96. Device said it's not supported.
[I] Attached Devices: 1: 0x1b7f; 2: 0x1b20; 3: 0x1b7e; 4: 0x0; 5: 0x0; 6: 0x0; 7: 0x0; 8: 0x0; 
[I] Setup finished for /dev/input/ckb2
[I] ckb3: Pairing id: ff ad 47 d6 bc 2b 6f 44 
[I] Setup finished for /dev/input/ckb3
```
ckb-next :
```
QObject::startTimer: Timers cannot have negative intervals
Path to settings: "/home/olivier/.config/ckb-next/ckb-next.conf"
KStatusNotifierWatcher not reachable
Scanning "/usr/local/libexec/ckb-next-animations/gradient"
Scanning "/usr/local/libexec/ckb-next-animations/heat"
Scanning "/usr/local/libexec/ckb-next-animations/invaders"
Scanning "/usr/local/libexec/ckb-next-animations/life"
Scanning "/usr/local/libexec/ckb-next-animations/mviz"
Scanning "/usr/local/libexec/ckb-next-animations/pinwheel"
Scanning "/usr/local/libexec/ckb-next-animations/pipe"
Scanning "/usr/local/libexec/ckb-next-animations/rain"
Scanning "/usr/local/libexec/ckb-next-animations/random"
Scanning "/usr/local/libexec/ckb-next-animations/ripple"
Scanning "/usr/local/libexec/ckb-next-animations/snake"
Scanning "/usr/local/libexec/ckb-next-animations/wave"
KStatusNotifierWatcher not reachable
Starting "/usr/local/libexec/ckb-next-animations/wave"
Starting "/usr/local/libexec/ckb-next-animations/wave"
Starting "/usr/local/libexec/ckb-next-animations/gradient"
Starting "/usr/local/libexec/ckb-next-animations/wave"
Starting "/usr/local/libexec/ckb-next-animations/gradient"
Creating PulseAudio context
Notify thread returning. Read 0 isOpen() true
Notify thread returning. Read 0 isOpen() true
Notify thread returning. Read 0 isOpen() true
Disconnected from PulseAudio 6
Disconnecting from X
XCB EWMH Thread Quitting...

```

**Comment by gilvbp** [comment]
> Hello, I have added some little thing :
> 
> * DPI 3 level indicator & Sniper mode (working like in official)
> * Battery indicator (under wheel) (green = charging, orange = low, red = critical) (like in official... but not sure for orange)
> * Profile switch now working
> * Some little cleaning
> 
> The DPI level working better is that in black color and not in animation DarkCoreRGBPro_DPI
> 
> Here, the newest build with -DCMAKE_BUILD_TYPE=Debug :
> 
> ckb-next-daemon :
> 
> ```
> ckb-next: Corsair RGB driver 0.5.0-unknown
> [I] Root controller ready at /dev/input/ckb0
> [I] Adding device with vendor = 0x1b1c, product = 0x1b7f, path = /dev/bus/usb/001/016, syspath = /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3
> [E] patchkeys (keymap_patch.c:151): Could not find profswitch in base keymap. Not patching profswitch for mouse.
> [I] Adding device with vendor = 0x1b1c, product = 0x1b20, path = /dev/bus/usb/001/003, syspath = /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-2
> [I] Connecting CORSAIR DARK CORE RGB PRO SE Gaming Dongle at /dev/input/ckb1
> [I] Claiming interfaces. name=CORSAIR DARK CORE RGB PRO SE Gaming Dongle, firmware=0204, ep_str= 5
> [I] ckb1: Claiming 5 interfaces
> [I] Connecting Corsair STRAFE RGB Gaming Keyboard at /dev/input/ckb2
> [I] Claiming interfaces. name=Corsair STRAFE RGB Gaming Keyboard, firmware=0308, ep_str= 2
> [I] ckb2: Claiming 2 interfaces
> [I] Starting input thread for /dev/input/ckb1
> [I] Starting input thread for /dev/input/ckb2
> [I] Found EP 0x82 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.0/ep_82
> [I] Found EP 0x81 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-2/1-2:1.0/ep_81
> [I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-2/1-2:1.0/ep_81 has wMaxPacketSize 64
> [I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.0/ep_82 has wMaxPacketSize 64
> [I] Found EP 0x84 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.1/ep_84
> [I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.1/ep_84 has wMaxPacketSize 64
> [I] Found EP 0x83 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.2/ep_83
> [I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.2/ep_83 has wMaxPacketSize 64
> [I] Found EP 0x81 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.3/ep_81
> [I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.3/ep_81 has wMaxPacketSize 64
> [I] Found EP 0x85 at /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.4/ep_85
> [I] Endpoint path /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.1/usb1/1-1/1-1.3/1-1.3:1.4/ep_85 has wMaxPacketSize 64
> [I] Received identification packet: 0e 01 00 00 01 01 00 01 08 03 14 00 1c 1b 20 1b 01 01 01 08 c0 ff 40 01 03 00 00 00 02 44 01 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
> [W] bragi_get_property (bragi_common.c:13): Failed to get property 0x15. Device said it's not supported.
> [W] bragi_get_property (bragi_common.c:13): Failed to get property 0x16. Device said it's not supported.
> [W] bragi_get_property (bragi_common.c:13): Failed to get property 0x96. Device said it's not supported.
> [I] ckb1: Pairing id: ff ad 47 d6 bc 2b 6f 44 
> [I] Found new bragi subdevice 1
> [I] ckb3: Subkb vendor: 0x1b1c, product: 0x1b7e
> [E] patchkeys (keymap_patch.c:151): Could not find profswitch in base keymap. Not patching profswitch for mouse.
> [I] Setup finished for /dev/input/ckb1
> [W] bragi_get_property (bragi_common.c:13): Failed to get property 0x15. Device said it's not supported.
> [W] bragi_get_property (bragi_common.c:13): Failed to get property 0x16. Device said it's not supported.
> [W] bragi_get_property (bragi_common.c:13): Failed to get property 0x96. Device said it's not supported.
> [I] Attached Devices: 1: 0x1b7f; 2: 0x1b20; 3: 0x1b7e; 4: 0x0; 5: 0x0; 6: 0x0; 7: 0x0; 8: 0x0; 
> [I] Setup finished for /dev/input/ckb2
> [I] ckb3: Pairing id: ff ad 47 d6 bc 2b 6f 44 
> [I] Setup finished for /dev/input/ckb3
> ```
> 
> ckb-next :
> 
> ```
> QObject::startTimer: Timers cannot have negative intervals
> Path to settings: "/home/olivier/.config/ckb-next/ckb-next.conf"
> KStatusNotifierWatcher not reachable
> Scanning "/usr/local/libexec/ckb-next-animations/gradient"
> Scanning "/usr/local/libexec/ckb-next-animations/heat"
> Scanning "/usr/local/libexec/ckb-next-animations/invaders"
> Scanning "/usr/local/libexec/ckb-next-animations/life"
> Scanning "/usr/local/libexec/ckb-next-animations/mviz"
> Scanning "/usr/local/libexec/ckb-next-animations/pinwheel"
> Scanning "/usr/local/libexec/ckb-next-animations/pipe"
> Scanning "/usr/local/libexec/ckb-next-animations/rain"
> Scanning "/usr/local/libexec/ckb-next-animations/random"
> Scanning "/usr/local/libexec/ckb-next-animations/ripple"
> Scanning "/usr/local/libexec/ckb-next-animations/snake"
> Scanning "/usr/local/libexec/ckb-next-animations/wave"
> KStatusNotifierWatcher not reachable
> Starting "/usr/local/libexec/ckb-next-animations/wave"
> Starting "/usr/local/libexec/ckb-next-animations/wave"
> Starting "/usr/local/libexec/ckb-next-animations/gradient"
> Starting "/usr/local/libexec/ckb-next-animations/wave"
> Starting "/usr/local/libexec/ckb-next-animations/gradient"
> Creating PulseAudio context
> Notify thread returning. Read 0 isOpen() true
> Notify thread returning. Read 0 isOpen() true
> Notify thread returning. Read 0 isOpen() true
> Disconnected from PulseAudio 6
> Disconnecting from X
> XCB EWMH Thread Quitting...
> ```

Hi @Clofriwen, great job!
I have found one problem, the wheel button is changing the DPI mode. (sniper mode -> first dpi)

**Comment by tatokis** [comment]
Since my comments seem to have disappeared, why are dpiwX being used instead of dpiX? I'd like to avoid making new key definitions unless absolutely necessary.

**Comment by Clofriwen** [comment]
> Since my comments seem to have disappeared, why are dpiwX being used instead of dpiX? I'd like to avoid making new key definitions unless absolutely necessary.

@tatokis because dpiX already used, and on Dark Core RGB Pro, doesn't work like on iCUE (i have tested).
if i use dpiX, all dpi dot will get same color for 1sec et back to asked (and that at every change)


> Hi @Clofriwen, great job! I have found one problem, the wheel button is changing the DPI mode inside Performance TAB and only works ok there. (sniper mode -> another dpi ) Could you help? Thanks

@gilvbp you can change wheel button in Binding Tab as snipper mode

**Comment by gilvbp** [comment]
> 

@Clofriwen Thanks!

**Review by Clofriwen — COMMENTED** [review]
Request is done

**Review by Clofriwen — COMMENTED** [review]
Can see DEBUG result in this message :
https://github.com/ckb-next/ckb-next/pull/855#issuecomment-1179599620

**Comment by gilvbp** [comment]
> Can see DEBUG result in this message : [#855 (comment)](https://github.com/ckb-next/ckb-next/pull/855#issuecomment-1179599620)

Hi @Clofriwen did you resolve the wheel conflict? I tested and still not working.

**Comment by Clofriwen** [comment]
@gilvbp Hi, i don't see where is problem :
- in snipper mode, i push and dpi decrase, and come back at my release
- in cycle up/cycle down that change everytime i push it, and stay at my release

Did you correctly set it ? (you have to apply for try it)
Capture d’écran_2022-08-20_20-06-01

**Comment by gilvbp** [comment]
> @gilvbp Hi, i don't see where is problem :
> 
> * in snipper mode, i push and dpi decrase, and come back at my release
> * in cycle up/cycle down that change everytime i push it, and stay at my release
> 
> Did you correctly set it ? (you have to apply for try it) Capture d’écran_2022-08-20_20-06-01

@Clofriwen  Hi there! In the version ckb-next-0.5.0.tar.gz that you had provided the wheel button I can use to scroll pages fine, with this version I can´t, only works inside ckb-net program. (changing the first dpi to sniper mode)
print1

**Comment by Clofriwen** [comment]
For button, that only to do be detect, for effect, that not in part i have changed...

You can change you DPI with dpi button ? What is the work of middle-bottom button (Binding) ? Button work outside CKB ? If that work one function, well, button is ok, and problem is something else.

the binding is important for 3 buttons : 2 on side, and middle-bottom, and binding can be set in binding tab and by clicking on concerned button.

The other thing is binding button to an work, and that work, it's an user choose.
The DPI swap/change isn't an "Dark Core RGB Pro" part, i have only to set for that device min & max dpi.

**Comment by gilvbp** [comment]
Yes, dpi button works fine.
That's so strange because wheel button appears ok at binding window ( up & down). Did you merge the last ckb-next code with yours version, right? Maybe it's some conflict, or the main version breaked this functionality.

**Comment by colbiedison** [comment]
@gilvbp I just got a Dark Core RGB Pro and I tried this fork. I ran into the same issue where scrolling down would switch to Sniper mode and scrolling up would switch it back. I found a fix and made [a PR](https://github.com/Clofriwen/ckb-next/pull/1) to @Clofriwen 's fork.

I attempted merging the latest master but that leads to another issue: the Binding tab not rendering a new frame, until I resize the window or focus another one and come back. I'm not sure if it's caused by commits in this PR; I don't have any other Corsair devices to be able to get to that menu on the master branch.

**Comment by gilvbp** [comment]
@colbiedison I have tested your version, and it seems that's all good. I didn't have any issues with windows or focus on the binding tab.

As you saw in my screenshot I have two devices. If I catch any problem I will let you know.

Thanks for helping and great job!

@Clofriwen  please check this merge request https://github.com/Clofriwen/ckb-next/pull/1

**Comment by Clofriwen** [comment]
@gilvbp Hello, i'm actualy under Pirating (hacking) attack on all my internet accound (after get an virus on a windows PC that grabbed all account acces).

So, i'm reparing/preventing my account, after that i will merge with @colbiedison solution (if he write he's agree) and an conflicting correction.

i hope be able to do it this week-end

**Comment by Clofriwen** [comment]
Hello, i did an Conflict resolution and merged @colbiedison fix.

In test, DPI UP button can't be detected by daemon... probably an issue due to lastest mouse firmware update (5.0.41)

@gilvbp & @colbiedison can you test actual commit (with your mouse firmware version) ?

Thanks your.

**Comment by gilvbp** [comment]
@Clofriwen @colbiedison  I tested and everything seems to work. (DPI UP & DOWN it's OK)

**Comment by Clofriwen** [comment]
My DPI UP button problem is on Windows too, so, that from my mouse or firmware (I tried a forced update, that no change)

**Comment by colbiedison** [comment]
@Clofriwen I tested https://github.com/Clofriwen/ckb-next/commit/ab2096ea5a9d2c0bb41d09c3695f44f42b4cc734 (your conflict resolution merge) and everything seems to work fine except the Binding tab is still not rendering properly. Perhaps I should open a separate issue? I think it's probably related to https://github.com/ckb-next/ckb-next/commit/40c53ab0a9cdb0d962a4c6d637e330dd541a69de and not this fork.
The binding tab works normally on this branch before the merge.

**Comment by gilvbp** [comment]
@tatokis  Hi there, Can you do the merge? Thanks

**Comment by olokos** [comment]
I have tried the latest official source and after enabling experimental mode left and right click dont work anymore and the mouse is almost useless, without experimental mode back and forwards enter 1 or 2 instead.

With Clowifren's fork the mouse gets detected and even shows battery level in the deamon and all, most of the mouse features work fine with it. I'll stick to the fork until this is merged.

**Comment by nlynzaad** [comment]
using the fork all settings seems to work using Dark Core RGB Pro SE

**Comment by Clofriwen** [comment]
Resolved lastest conflict

**Comment by tatokis** [comment]
Thanks. I'll have a look once I'm done with the current release. There's a huge backlog, apologies.

**Pull request merged**
